marlowe-contracts/test/Spec/Marlowe/Contracts.hs

```diff
@@ -0,0 +1,395 @@
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE OverloadedStrings  #-}
+module Spec.Marlowe.Contracts
+    (tests)
+where
+
+import Language.Marlowe
+import Ledger.Ada
+import Ledger.Value
+import Marlowe.Contracts.Common
+import Marlowe.Contracts.Futures
+import Marlowe.Contracts.Options
+import Marlowe.Contracts.StructuredProducts
+import Marlowe.Contracts.Swap
+import Marlowe.Contracts.ZeroCouponBond
+import Test.Tasty
+import Test.Tasty.HUnit
+
+tests :: TestTree
+tests = testGroup "Marlowe Contract"
+  [ testCase "ZeroCouponBond" zeroCouponBondTest
+  , testCase "Swap Contract" swapContractTest
+  , testCase "Future Contract (repayment of initial margin)" futureNoChange
+  , testCase "Future Contract (without margin calls)" futureNoMarginCall
+  , testCase "Future Contract (with margins calls)" futureWithMarinCall
+  , testCase "American Call Option test" americanCallOptionTest
+  , testCase "American Call Option (exercised) test" americanCallOptionExercisedTest
+  , testCase "European Call Option test" europeanCallOptionTest
+  , testCase "European Call Option (exercised) test" europeanCallOptionExercisedTest
+  , testCase "Reverse Convertible test" reverseConvertibleTest
+  , testCase "Reverse Convertible (exercised) test" reverseConvertibleExercisedTest
+  ]
+
+w1Pk, w2Pk :: Party
+w1Pk = Role "party1"
+w2Pk = Role "party2"
+
+tokName :: TokenName
+tokName = "testcoin"
+
+tokSymbol :: CurrencySymbol
+tokSymbol = ""
+
+tok :: Token
+tok = Token tokSymbol tokName
+
+tokValueOf :: Integer -> Money
+tokValueOf = singleton tokSymbol tokName
+
+assertTotalPayments :: Party -> [Payment] -> Money -> Assertion
+assertTotalPayments p t x = assertBool "total payments to party" (totalPayments t == x)
+  where
+    totalPayments = mconcat . map (\(Payment _ _ a) -> a) . filter (\(Payment _ a _) -> a == Party p)
+
+assertNoWarnings :: [a] -> Assertion
+assertNoWarnings [] = pure ()
+assertNoWarnings t  = assertBool "No warnings" $ null t
+
+assertClose :: Contract -> Assertion
+assertClose = assertBool "Contract is in Close" . (Close==)
+
+step :: TransactionInput -> ([TransactionWarning], [Payment], State, Contract) -> Either TransactionError ([TransactionWarning], [Payment], State, Contract)
+step i (_,_,s,c) =
+  case computeTransaction i s c of
+    Error err                   -> Left err
+    TransactionOutput w p s' c' -> Right (w, p, s', c')
+
+-- |Zero-coupon Bond test
+zeroCouponBondTest :: IO ()
+zeroCouponBondTest =
+  let zcb =
+        zeroCouponBond
+          w1Pk
+          w2Pk
+          (Slot 100)
+          (Slot 200)
+          (Constant 75_000_000)
+          (Constant 90_000_000)
+          ada
+          Close
+   in either left right $
+        step (TransactionInput (0, 0)     [NormalInput $ IDeposit w1Pk w1Pk ada 75_000_000]) ([], [], emptyState 0, zcb)
+    >>= step (TransactionInput (100, 110) [NormalInput $ IDeposit w1Pk w2Pk ada 90_000_000])
```

Comment:
There is a `playTrace` function defined in https://github.com/input-output-hk/marlowe/blob/master/src/Language/Marlowe/Semantics.hs#L417, I didn't realise it wasn't in this repo. Would you mind copying it to `Language.Marlowe.Semantics` and using that instead of `Either`? `playTrace` is the one we verified several properties for, and will make all this code much simpler cause it just becomes pattern matching on `TransactionOutput`